target/linux/lantiq/patches-5.10/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch

```diff
@@ -0,0 +1,99 @@
+From 7c481d31cbe49704b6326ea46ba84c2be2c61d95 Mon Sep 17 00:00:00 2001
+From: Thomas Nixon <tom@tomn.co.uk>
+Date: Sat, 3 Jul 2021 15:57:03 +0100
+Subject: [PATCH] net: lantiq: enable baby jumbo frames on xrx200
+
+xrx200 MTU is reduced so that the maximum mtu setting works correctly.
+
+The maximum packet size accepted by the switch is increased to match the
+maximum mtu.
+
+Signed-off-by: Thomas Nixon <tom@tomn.co.uk>
+---
+ drivers/net/dsa/lantiq_gswip.c       | 26 +++++++++++++++++++++++---
+ drivers/net/ethernet/lantiq_xrx200.c |  4 +++-
+ 2 files changed, 26 insertions(+), 4 deletions(-)
+
+diff --git a/drivers/net/dsa/lantiq_gswip.c b/drivers/net/dsa/lantiq_gswip.c
+index 93c7fa1fd4cb..da127d6b535b 100644
+--- a/drivers/net/dsa/lantiq_gswip.c
++++ b/drivers/net/dsa/lantiq_gswip.c
+@@ -238,6 +238,9 @@
+ 
+ #define XRX200_GPHY_FW_ALIGN	(16 * 1024)
+ 
++/* max mtu including tag size */
++#define GSWIP_MAX_MTU	(1508 + 8)
++
+ struct gswip_hw_info {
+ 	int max_ports;
+ 	int cpu_port;
+@@ -851,9 +854,14 @@ static int gswip_setup(struct dsa_switch *ds)
+ 	gswip_switch_mask(priv, 0, GSWIP_PCE_PCTRL_0_INGRESS,
+ 			  GSWIP_PCE_PCTRL_0p(cpu_port));
+ 
+-	gswip_switch_mask(priv, 0, GSWIP_MAC_CTRL_2_MLEN,
+-			  GSWIP_MAC_CTRL_2p(cpu_port));
+-	gswip_switch_w(priv, VLAN_ETH_FRAME_LEN + 8, GSWIP_MAC_FLEN);
++	/* enable max mtu on all ports */
++	for (i = 0; i < priv->hw_info->max_ports; i++)
++		gswip_switch_mask(priv, 0, GSWIP_MAC_CTRL_2_MLEN,
++				  GSWIP_MAC_CTRL_2p(i));
++	/* includes header, special tag, and FCS */
++	gswip_switch_w(priv, VLAN_ETH_HLEN + GSWIP_MAX_MTU + ETH_FCS_LEN,
++		       GSWIP_MAC_FLEN);
++
+ 	gswip_switch_mask(priv, 0, GSWIP_BM_QUEUE_GCTRL_GL_MOD,
+ 			  GSWIP_BM_QUEUE_GCTRL);
+ 
+@@ -1426,6 +1434,16 @@ static int gswip_port_fdb_dump(struct dsa_switch *ds, int port,
+ 	return 0;
+ }
+ 
++static int gswip_port_max_mtu(struct dsa_switch *ds, int port)
++{
++	return GSWIP_MAX_MTU;
++}
++
++static int gswip_port_change_mtu(struct dsa_switch *ds, int port, int new_mtu)
++{
++	return 0;
```

Comment:
You should only write to `GSWIP_MAC_FLEN` in `.port_change_mtu`, instead of hardcoding it elsewhere.

Please consider upstreaming this too.

Reply:
Thanks for taking a look. My reason for not doing this is that there is only one FLEN register shared between all switch ports, which also needs to account for the DSA tag on the CPU port. Therefore when the MTU of a port is set to anything other than 1500, it is always going to be able to accept larger packets than intended.

Since this doesn't seem to have any negative effects, we might as well just set all ports to support the maximum size we want, and save some code in the process. Does this make sense, or am i being lazy?

I'll have a go and see how it looks anyway.

Reply:
You can take a look at how I implement this for MT7530 switch: torvalds/linux@9470174e7581e75a8ebd78964997314dfc2e706c, remember to test the edge cases!
And if the global `GSWIP_MAC_FLEN` register actually configures max RX length (MRU), you should enable `mtu_enforcement_ingress` in setup function: torvalds/linux@771c8901568dd8776a260aa93db41be88a60389e

Reply:
Thanks, that looks a bit cleaner than what i was thinking about -- I'll give it a go later.

Reply:
Latest push should be as per your suggestion, thanks for that.

The 5.4 and 5.10 patches now operate quite differently; I'm not exactly sure how is best to handle that. On the other hand, the 5.10 patch now doesn't really change anything if the user doesn't change the MTU, so it's a bit less scary.

Reply:
If the switch supports 9K jumbo frames, you should lift the limit.

Reply:
Sorry i meant to mention that; that was the highest i was able to test so far. I'll have another go later; it's a bit tricky, because the ethernet driver doesn't seem to work right with larger frames.

Reply:
Updated to use the max packet length from the DM200 GPL dump. I can't test above 9K, but the switch still passes 9K packets when set to the maximum (with the ethernet driver modified to allow higher MTUs), so it should be fine.

In practice it's always limited by the max MTU of the ethernet driver, so this is only really relevant for someone working on modifying the ethernet driver to support larger packets.

Reply:
Does the CPU port always need `GSWIP_MAC_CTRL_2_MLEN`, even if all user ports MTU <= 1492?
If it does not, you can simplify the function
```c
if (port == cpu_port) {
	new_mtu += 8;
	gswip_switch_w(priv, VLAN_ETH_HLEN + new_mtu + ETH_FCS_LEN,
		       GSWIP_MAC_FLEN);
}

if (new_mtu > ETH_DATA_LEN)
	gswip_switch_mask(priv, 0, GSWIP_MAC_CTRL_2_MLEN,
			  GSWIP_MAC_CTRL_2p(port));
else
	gswip_switch_mask(priv, GSWIP_MAC_CTRL_2_MLEN, 0,
			  GSWIP_MAC_CTRL_2p(port));
```

Reply:
Good point, I've made that change.

Also changed `>` to `!=`, because MLEN/FLEN also work for reducing the maximum accepted  packet size.